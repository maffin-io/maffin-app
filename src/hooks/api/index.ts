import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
