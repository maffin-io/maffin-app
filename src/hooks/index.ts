export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
