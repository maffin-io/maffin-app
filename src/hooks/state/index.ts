export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
