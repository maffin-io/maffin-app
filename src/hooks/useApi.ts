import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
