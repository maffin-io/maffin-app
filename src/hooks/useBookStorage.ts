import React from 'react';

import useGapiClient from '@/hooks/useGapiClient';
import type BookStorage from '@/lib/storage/BookStorage';
import {
  GDriveBookStorage,
  DemoBookStorage,
} from '@/lib/storage';
import useSession from './useSession';

type UseBookStorageReturn = {
  storage: BookStorage | null,
};

export default function useBookStorage(): UseBookStorageReturn {
  const [isGapiLoaded] = useGapiClient();
  const [storage, setStorage] = React.useState<BookStorage | null>(null);
  const { roles } = useSession();

  const [error, setError] = React.useState<Error | null>(null);
  if (error) {
    throw error;
  }

  React.useEffect(() => {
    async function load() {
      let instance: BookStorage = new DemoBookStorage();

      if (roles.isPremium) {
        instance = new GDriveBookStorage(window.gapi.client);
      }

      try {
        await instance.initStorage();
        setStorage(instance);
      } catch (e) {
        setError(e as Error);
      }
    }

    if (isGapiLoaded) {
      load();
    }
  }, [isGapiLoaded, roles.isPremium]);

  return { storage };
}
