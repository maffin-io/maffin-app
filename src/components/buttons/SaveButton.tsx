import React from 'react';
import { BiCloudUpload, BiLoader, BiWifiOff } from 'react-icons/bi';
import { useQuery } from '@tanstack/react-query';
import classNames from 'classnames';

import { DataSourceContext } from '@/hooks';
import { IS_PAID_PLAN } from '@/helpers/env';
import { useOnline } from '@/hooks/state';
import { UpgradeTooltip } from '@/components/tooltips';

export default function SaveButton(): JSX.Element {
  const { data: isSaving } = useQuery({
    queryKey: ['state', 'isSaving'],
    queryFn: () => false,
    networkMode: 'always',
  });
  const { isLoaded, save } = React.useContext(DataSourceContext);
  const { isOnline } = useOnline();

  if (!isLoaded) {
    return (
      <button
        type="button"
        className="btn btn-primary"
      >
        ...
      </button>
    );
  }

  return (
    <>
      <button
        id="save-button"
        type="button"
        className={classNames(
          'btn btn-primary',
          {
            'btn-danger': !isOnline,
          },
        )}
        disabled={isSaving || !IS_PAID_PLAN || !isOnline}
        data-tooltip-id="upgrade-tooltip"
        onClick={async () => {
          await save();
        }}
      >
        {
          (
            isSaving
            && isOnline
            && (
            <>
              <BiLoader className="mr-1 animate-spin" />
              <span>Saving...</span>
            </>
            )
          ) || (
            !isOnline
            && (
              <>
                <BiWifiOff className="mr-1" />
                <span>Save</span>
              </>
            )
          ) || (
            <>
              <BiCloudUpload className="mr-1" />
              <span>Save</span>
            </>
          )
        }
      </button>
      {
        !IS_PAID_PLAN
        && <UpgradeTooltip />
      }
    </>
  );
}
