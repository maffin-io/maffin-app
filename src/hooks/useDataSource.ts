import React from 'react';
import { DataSource } from 'typeorm';
import initSqlJs from 'sql.js';
import { mutate } from 'swr';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/Stocker';
import useBookStorage from '@/hooks/useBookStorage';
import { importBook as importGnucashBook } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { isStaging } from '@/helpers/env';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        await loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  mutate('/state/save', true, { revalidate: false });
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  await storage.save(rawBook);
  mutate('/state/save', false, { revalidate: false });
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const rawBook = await importGnucashBook(parsedData);
  await DATASOURCE.sqljsManager.loadDatabase(rawBook);

  DATASOURCE.options.extra.queryClient.refetchQueries({
    type: 'all',
  });

  await loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  if (!isStaging()) {
    try {
      // We have to await because we use Price.upsert. Once we correct this,
      // saving Price already updates the local cache so we don't need await
      await insertTodayPrices();
    } catch (e) {
      console.warn(`Retrieving live prices failed ${e}`);
    }
  }
}
