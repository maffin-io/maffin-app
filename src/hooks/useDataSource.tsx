import React from 'react';
import { DataSource, QueryFailedError } from 'typeorm';
import initSqlJs from 'sql.js';
import { Settings } from 'luxon';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/prices';
import useBookStorage from '@/hooks/useBookStorage';
import { migrate as migrateFromGnucash } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';
import { MaffinError, StorageError } from '@/helpers/errors';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], true);
  await DATASOURCE.query('VACUUM');
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  try {
    await storage.save(rawBook);
  } catch (e) {
    (e as MaffinError).show();
  } finally {
    DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], false);
  }
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  if (process.env.NEXT_PUBLIC_ENV !== 'master') {
    Settings.now = () => Date.now();
  }
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const previousData = DATASOURCE.sqljsManager.exportDatabase();
  await DATASOURCE.sqljsManager.loadDatabase(parsedData);
  try {
    // Sadly loadDatabase doesn't check that the file is valid so we call this to trigger
    // errors if any
    await DATASOURCE.query('VACUUM');
  } catch (e) {
    let code = 'UNKNOWN';
    if (e instanceof QueryFailedError) {
      code = 'INVALID_FILE';
    }
    new StorageError((e as Error).message, code).show();
    await DATASOURCE.sqljsManager.loadDatabase(previousData);
    return;
  }

  // We probably should do this conditionally only for when it is a gnucash file
  await migrateFromGnucash();
  DATASOURCE.options.extra.queryClient.refetchQueries({
    queryKey: ['api'],
    type: 'all',
  });

  loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  try {
    // We have to await because we use Price.upsert. Once we correct this,
    // saving Price already updates the local cache so we don't need await
    await insertTodayPrices();
  } catch (e) {
    console.warn(`Retrieving live prices failed ${e}`);
  }
}
