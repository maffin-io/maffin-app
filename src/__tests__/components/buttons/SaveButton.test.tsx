import React from 'react';
import {
  act,
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { SWRConfig, mutate } from 'swr';

import SaveButton from '@/components/buttons/SaveButton';
import type { UseDataSourceReturn } from '@/hooks';
import * as dataSourceHooks from '@/hooks/useDataSource';

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('@/hooks/useBookStorage', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useBookStorage'),
}));

describe('SaveButton', () => {
  it('loads while unavailable datasource', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue(
      { isLoaded: false } as UseDataSourceReturn,
    );
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SaveButton />
      </SWRConfig>,
    );

    await screen.findByText('...');
    expect(container).toMatchSnapshot();
  });

  it('renders as expected when datasource ready and not saving', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue(
      { isLoaded: true } as UseDataSourceReturn,
    );
    const { container } = render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SaveButton />
      </SWRConfig>,
    );

    await screen.findByText('Save');
    expect(container).toMatchSnapshot();
  });

  it('renders as expected when datasource ready and saving', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue(
      { isLoaded: true } as UseDataSourceReturn,
    );
    const { container } = render(<SaveButton />);

    act(() => {
      mutate('/state/save', true, { revalidate: false });
    });

    expect(screen.getByRole('button')).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('calls datasource save on click', async () => {
    const mockSave = jest.fn();
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue(
      {
        isLoaded: true,
        save: mockSave as Function,
      } as UseDataSourceReturn,
    );
    render(
      <SWRConfig value={{ provider: () => new Map() }}>
        <SaveButton />
      </SWRConfig>,
    );

    const button = screen.getByText('Save');

    fireEvent.click(button);

    await waitFor(() => expect(mockSave).toBeCalledTimes(1));
  });
});
