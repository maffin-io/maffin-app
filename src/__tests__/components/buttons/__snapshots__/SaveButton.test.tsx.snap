// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveButton loads while unavailable datasource 1`] = `
<div>
  <button
    class="btn btn-primary"
    type="button"
  >
    ...
  </button>
</div>
`;

exports[`SaveButton renders as expected when datasource ready and not saving 1`] = `
<div>
  <button
    class="btn btn-primary"
    id="save-button"
    type="button"
  >
    <svg
      class="mr-1"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13 19v-4h3l-4-5-4 5h3v4z"
      />
      <path
        d="M7 19h2v-2H7c-1.654 0-3-1.346-3-3 0-1.404 1.199-2.756 2.673-3.015l.581-.102.192-.558C8.149 8.274 9.895 7 12 7c2.757 0 5 2.243 5 5v1h1c1.103 0 2 .897 2 2s-.897 2-2 2h-3v2h3c2.206 0 4-1.794 4-4a4.01 4.01 0 0 0-3.056-3.888C18.507 7.67 15.56 5 12 5 9.244 5 6.85 6.611 5.757 9.15 3.609 9.792 2 11.82 2 14c0 2.757 2.243 5 5 5z"
      />
    </svg>
    <span>
      Save
    </span>
  </button>
</div>
`;

exports[`SaveButton renders as expected when datasource ready and saving 1`] = `
<div>
  <button
    class="btn btn-primary"
    disabled=""
    id="save-button"
    type="button"
  >
    <svg
      class="mr-1 animate-spin"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M2 11h5v2H2zm15 0h5v2h-5zm-6 6h2v5h-2zm0-15h2v5h-2zM4.222 5.636l1.414-1.414 3.536 3.536-1.414 1.414zm15.556 12.728-1.414 1.414-3.536-3.536 1.414-1.414zm-12.02-3.536 1.414 1.414-3.536 3.536-1.414-1.414zm7.07-7.071 3.536-3.535 1.414 1.415-3.536 3.535z"
      />
    </svg>
    <span>
      Saving...
    </span>
  </button>
</div>
`;
